### DISCUSSION TITLE
Better docs for creating a distributable package

### DISCUSSION THREAD

**Discussion by ghost** [IndiePython/nodezator#89] [Concerns/constructive criticism]
It took me a lot of trial and error to settle on my current structure:
```
[...]/dotenv_ndz$ tree
.
├── dotenv_ndz
│   └── dotenv
│       └── get_var
│           ├── __main__.py
│           └── __pycache__
│               ├── __main__.cpython-310.pyc
│               └── __main__.cpython-311.pyc
├── dotenv_ndz.egg-info
│   ├── dependency_links.txt
│   ├── PKG-INFO
│   ├── requires.txt
│   ├── SOURCES.txt
│   └── top_level.txt
├── pyproject.toml
└── requirements.txt
```

as this supports editable installs (and hopefully) it's PyPi ready. (Plus now this package can be imported, as without that top-level folder it could only be done from the parent directory for some reason, thus Nodezator didn't see this package.)

Could I improve this? Am I overlooking smt.?

**Comment by KennedyRichard** [comment]
I can't actually determine whether this can be improved or whether something is being overlooked without actually inspecting the metadata files and configuration. But don't worry, this shouldn't be needed.

Also, I'm not a Python packaging specialist either (and I doubt most Python users are), but I'll be glad to help you as much as I can.

First, you may already know, but there's a template/example PyPI-ready installable node pack here: https://github.com/KennedyRichard/mynodepack. This template actually works and you can even install it with `pip install mynodepack` and try it.

Here are a few additional pointers taking it into account:

- Turning your node pack into a PyPI-ready shouldn't require many additional files:
  - Just the folder containing your node pack
  - And a few metadata files
    - setup.cfg
    - pyproject.toml
    - README.md (not actually required but nice to have)
  - not even the .gitignore file is required for packaging, it is there just to help working with .git
  - you should replace the UNLICENSE file by a LICENSE file of your chosen license (unless you want your code in the public domain)
- the README file of the example node pack has a detailed step-by-step guide describing the entire packaging process
- there may be other ways to go about packaging, but they are not necessarily better; for instance, nowadays it is possible to ditch the setup.cfg file altogether and only include metadata info in the pyproject.toml file, but I know of a Python packaging specialist that still prefers using setup.cfg for metadata; in other words, the method I describe/use in the template should cover everything one needs in terms of functionality, including settings for automatically downloading and installing dependencies when user of your node pack install it via pip.

Finally, in case you are not aware. PyPI has actually a duplicate service for testing purposes called TestPyPI. You can use it in the very same way you'd use PyPI and even install packages from it in the same way. This is great because you can upload your package there as many times as you'd like as you fine-tune your settings or source code and only after ensuring it is working correctly you can upload the package to the regular PyPI server.

This process is also described in the README of the template package I mentioned above.

If you still have any questions, don't hesitate to ask.

Sorry for taking so long to answer. As you know, I've been doing a lot of work in and outside the Indie Python/Nodezator projects, including the recent patch I released to fix your issue regarding the overflowing dialog.

Finally, in the title of this discussion, you suggest the need for better docs for creating a distributable package, but you didn't actually mention any docs, only your own trial and error. Could you please mention which docs in specific you are referring to that you think should be improved?

Could it be, perhaps, that you weren't aware of the existence of the example/template node pack I mentioned, or of the chapter in the manual that mentions it (https://manual.nodezator.com/ch-distributing-nodes.html)? This is a genuine question and you don't need to worry at all if you didn't know. If this is the case, it just means that it may be my fault for not making such materials more noticeable in the manual and other relevant spots.

I just wanted to know this because it is what you claimed in the title, so I didn't want to finish this reply without addressing it.

As always, let me know if you need anything else. I'll be glad to help if it is within my abilities.

Peace.
